src/std/misc/string.ss

```diff
@@ -77,3 +79,96 @@ package: std/misc
+;;   new    non-empty
+;;   count  non-zero, number of replacements (-1 means no limit)
+(def (subst-helper-empty-old str new count)
+  (declare (fixnum))
+  (def len-str (string-length str))
+  (if (= count 1)
+    (string-append new str)         ; add 'new' and leave procedure
+    (call-with-output-string
+     (lambda (port)
+       (write-string new port)      ; 'count' > 1, add 'new' before the first character
+       (let ((stop (1- len-str))
+             (count (if (or (negative? count) (> count len-str))
+		      (1+ len-str)  ; the maximal number of replacements is len + 1
+		      count)))
+	 (let loop ((i 0)
+		    (matches 1))    ; 1 because 'new' was already added once
+	   (cond
+	    ((= matches count)
+	     (write-string new port)
+	     (write-substring str i len-str port))
+	    ((= i stop)
+	     (unless (zero? i) (write-string new port))
+	     (write-char (string-ref str i) port)
+	     (write-string new port))
+	    (else
+	     (unless (zero? i) (write-string new port))
+	     (write-char (string-ref str i) port)
+	     (loop (1+ i) (1+ matches))))))))))
+
+
+;; string-subst helper which handles the case that the argument 'old' is a non-empty string.
+;;   str    non-empty
+;;   old    non-empty
+;;   new    can be empty
+;;   count  non-zero, number of replacements (-1 means no limit)
+(def (subst-helper-nonempty-old str old new count)
+  (declare (fixnum))
+  (def len-str (string-length str))
+  (def size-old (1- (string-length old)))
+  (def size-str (1- (string-length str)))
+  (call-with-output-string
+   (lambda (port)
+     (let loop ((i 0)       ; position in str
+		(matches 0)
+		(last 0)    ; position after last match in str
+		(j 0))      ; position in old
+       (cond
+	((= matches count)  ; stop, limit reached
+	 (write-substring str i len-str port))
+	((= i size-str)     ; stop, end of str
+	 (if (and (eq? (string-ref str i) (string-ref old j))
+                  (= j size-old))
+	   (write-string new port)
+	   (write-substring str last len-str port)))
+	(else
+	 (if (eq? (string-ref str i) (string-ref old j))
+           (if (= j size-old)                        ; match of old in str
+	     (begin
+	       (write-string new port)
+	       (loop (1+ i) (1+ matches) (1+ i) 0))
+	     (loop (1+ i) matches last (1+ j)))      ; char equal, not yet a match
+	   (begin
+	     (write-substring str last (1+ i) port)  ; no match, continue search
+	     (loop (1+ i) matches (1+ i) 0)))))))))
+
+
+;; In the copy of str replace the string old with string new.
+;; The procedure accepts only a fixnum or #f for count.
+;;   count > 0   limit replacements
+;;   count #f    no limit
+;;   count <= 0  return input
+;;
+;; Example:
+;;  (string-subst "abc" "b" "_") => "a_c"
+;;  (string-subst "abc" "" "_")  => "_a_b_c_"
+(def (string-subst str old new count: (count #f))
+  (declare (fixnum))
+  (when (true? count)
+    (error "Illegal argument; count must be a fixnum or #f, got #t"))
```

Comment:
why this explicit check for `#t`? I don't think we need this.
If you want to add a check for the type, add an explicit check for `fixnum?` or `#f`.

Reply:
Not that it is incorrect as a generic check, if you pass a symbol `'foo` then it won't be caught.

Reply:
So in short, either there should be no check at all or a check that validates it's a fixnum or `#f`.

Reply:
Can you update the error message too? It's misleading (there is this "got #t" thingie).

Reply:
ups